File: HexChat/duplicates/statuscolor.py
import hexchat

__module_name__ = 'statuscolor'
__module_author__ = 'TingPing'
__module_version__ = '0'
__module_description__ = 'Colors nicks based on user mode'
# This is just a direct port of LifeIsPain's statuscolor.pl script.

events = ('Channel Message',
		'Channel Action',
		'Channel Msg Hilight',
		'Channel Action Hilight',
		'Your Message',
		'Your Action')

# TODO: update these for hexchats default theme
modes = {'+':'24',
		'%':'28',
		'@':'19',
		'&':'21',
		'~':'22'}

edited = False

def msg_cb(word, word_eol, event, attrs):
	global edited
	if edited or len(word) < 3 or not word[2] in modes:
		return hexchat.EAT_NONE

	color = modes[word[2]]
	nick = '\003{}{}\00399'.format(color, hexchat.strip(word[0]))
	word = [(word[i] if len(word) > i else '') for i in range(4)]

	edited = True
	hexchat.emit_print(event, nick, word[1], word[2], word[3], time=attrs.time)
	edited = False

	return hexchat.EAT_ALL

for event in events:
	hexchat.hook_print_attrs(event, msg_cb, event, hexchat.PRI_HIGH)

File: HexChat/statuscolor.lua
-- SPDX-License-Identifier: MIT
hexchat.register('StatusColor', '1', 'Color nicks based upon user modes')

-- TODO: Update these to reflect default theme
local MODES = {
	['+'] = '24',
	['%'] = '28',
	['@'] = '19',
	['&'] = '21',
	['~'] = '22',
}

local edited = false
local function on_message (args, attrs, event)
	if edited then
		return hexchat.EAT_NONE
	end

	local color = MODES[args[3]]
	if not color then
		return hexchat.EAT_NONE
	end

	-- In > 2.12.3 we need to be explicit about color changes
	if event:sub(-7, -1) == 'Hilight' then
		hexchat.command('gui color 3')
	elseif event:sub(1, 4) ~= 'Your' then
		hexchat.command('gui color 2')
	end

	edited = true
	args[1] = '\003' .. color .. hexchat.strip(args[1]) .. '\00399'
	args[3] = '\003' .. color .. args[3] .. '\00399'
	hexchat.emit_print_attrs(attrs, event, unpack(args))
	edited = false

	return hexchat.EAT_ALL
end

for _, event in pairs({
	'Channel Message',
	'Channel Action',
	'Channel Msg Hilight',
	'Channel Action Hilight',
	'Your Message',
	'Your Action'
}) do
	hexchat.hook_print_attrs(event, function (args, attrs)
		return on_message(args, attrs, event)
	end, hexchat.PRI_LOW)
end
